WordPress/openverse
# Derive document sensitivity markings from filtered index

## Fixes



Fixes #1199 by @sarayourfriend 

## Description


This PR includes significant additions to the test factories and fixtures in order to facilitate Elasticsearch document creation. I chose this approach over ElasticMock for the following reasons:

1. It allows us to avoid mocking altogether (great for tests generally)
2. It exercises our _actual_ indexes, not fake ones
3. It still follows the general test-data hygiene rules of removing test data after each test case

The last one is the source of some abstraction that is not necessarily easy to follow (in that the parts aren't explicitly connected), but at least somewhat tied together via the `CREATED_BY_FIXTURE_MARKER`. The usage of pytest's `request` fixture is also potentially obscure but is quite common in pytest plugins of any significant complexity. I learned that specific usage (how to look for a specific mark) reading `pytest-django`'s code.

I also added a new `media_type_config` fixture that provides a dataclass containing most of the relevant disparate bits for different media types necessary for many of our tests. This made it easier to update tests that were previously using _very_ naive and shallow stubs which broke due to the new requirements. Luckily, the changes to those tests are easier to understand and will continue to benefit from additional improvements to the fixture classes. We'll be able to continue cleaning up tests with this fixture in the future in separate PRs, but not this one. In this one I tried to touch only the tests that actually needed changes.

Anyway, with that in mind, here is a summary of the significant changes:

- Enable Elasticsearch document creation in model factories
  - I needed to fix the Postgres ID creation for this because it is also used to set the document ID in Elasticsearch. The query for the latest ID needs to go to Elasticsearch because in tests the models are not able to query the existing data (it seems).
- Add `unstable__sensitivity` to the result serializer
  - This is marked as `unstable` and will be stabilised alongside the new query parameter after testing
- Remove ``request`` parameter from `search_controller.search` and `search_controller.related_media`. This parameter was not being used and made tests slightly more annoying to write as you had to either pass a real request or `None`. The first requires calling a new fixture and the other is obscure and unclear.
- Update many unit tests to accommodate the new `search_controller.search` interface (made easier by the improved fixtures and factories).

This PR includes a deviation from the original plan. Specifically, in the course of writing this PR and testing it, I discovered that we do not actually have _any_ documents marked mature in our production dataset that are not marked mature due to confirmed user reports. That's to say: we have no provider-supplied maturity. I left the constant in there along with a big comment in the new serializer method explaining why that is. Trying to accommodate that in this PR would significantly increase the complexity because we need to be able to efficiently (read: no duplicate queries) read from the database _and_ ES, regardless of the serialization context. It turns out that's not that hard for our actual runtime, but many many many many many tests assume the ability to bypass Elasticsearch (probably because we did not have easy Elasticsearch fixture creation). We need to put off that feature and luckily we don't need it because simply put, there are no results that fall into that case.

## Testing Instructions

Review the new unit tests. There are significant changes to the test factories to support these tests and they integrate directly with the live Elasticsearch instance.

Additionally, test by making queries. You must enable `ENABLE_FILTERED_INDEX_QUERIES` for the new parameter to work. Refer to `ingestion_server/static/mock_sensitive_terms.txt` for the terms used to generate the local filtered indexes (these terms are _not_ sensitive and are only for testing).

You should observe that requests that do not include sensitive results should always have `[]` for `unstable__sensitivity`.

You must create a confirmed mature report for results to test the "user reported sensitive" results.

You should query for one of the mock sensitive terms directly while including sensitive results and confirm that the documents appropriately include the sensitive text marking in `unstable__sensitivity`. Try a combination of the two as well, (i.e., create a confirmed report for a "bird" result locally). The new field should have both text and user-reported sensitivity indications.

## Checklist


- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
